tests/unit/s2n_tls13_server_cert_selection_test.c

```diff
@@ -0,0 +1,226 @@
+            EXPECT_SUCCESS(s2n_read_test_pem(test.private_key, private_key, S2N_MAX_TEST_PEM_SIZE));
+
+            struct s2n_config *server_config;
+            EXPECT_NOT_NULL(server_config = s2n_config_new());
+
+            struct s2n_cert_chain_and_key *default_cert;
+            EXPECT_NOT_NULL(default_cert = s2n_cert_chain_and_key_new());
+            EXPECT_SUCCESS(s2n_cert_chain_and_key_load_pem(default_cert, cert_chain, private_key));
+            EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(server_config, default_cert));
+
+            struct s2n_connection *server_conn;
+            EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, server_config));
+
+            /* test that our_chain_and_key is populated */
+            EXPECT_NULL(server_conn->handshake_params.our_chain_and_key);
+
+            struct s2n_signature_scheme sig_scheme_out = {0};
+            EXPECT_SUCCESS(s2n_choose_tls13_sig_scheme_and_set_cert(server_conn, &default_client_sig_scheme_pref_list, &sig_scheme_out));
+            EXPECT_EQUAL(sig_scheme_out.iana_value, test.iana_value);
+
+            EXPECT_SUCCESS(s2n_connection_free(server_conn));
+            EXPECT_SUCCESS(s2n_cert_chain_and_key_free(default_cert));
+            EXPECT_SUCCESS(s2n_config_free(server_config));
+
+            free(private_key);
+            free(cert_chain);
+        }
+    }
+
+    /* matching on empty client sig scheme list */
+    {
+        struct s2n_sig_scheme_list empty_client_sig_scheme_pref_list = {0};
+
+        char *cert_chain = NULL;
+        char *private_key = NULL;
+        EXPECT_NOT_NULL(cert_chain = malloc(S2N_MAX_TEST_PEM_SIZE));
+        EXPECT_NOT_NULL(private_key = malloc(S2N_MAX_TEST_PEM_SIZE));
+
+        EXPECT_SUCCESS(s2n_read_test_pem(S2N_DEFAULT_TEST_CERT_CHAIN, cert_chain, S2N_MAX_TEST_PEM_SIZE));
+        EXPECT_SUCCESS(s2n_read_test_pem(S2N_DEFAULT_TEST_PRIVATE_KEY, private_key, S2N_MAX_TEST_PEM_SIZE));
+
+        struct s2n_config *server_config;
+        EXPECT_NOT_NULL(server_config = s2n_config_new());
+
+        struct s2n_cert_chain_and_key *default_cert;
+        EXPECT_NOT_NULL(default_cert = s2n_cert_chain_and_key_new());
+        EXPECT_SUCCESS(s2n_cert_chain_and_key_load_pem(default_cert, cert_chain, private_key));
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(server_config, default_cert));
+
+        struct s2n_connection *server_conn;
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, server_config));
+
+        EXPECT_NULL(server_conn->handshake_params.our_chain_and_key);
+
+        struct s2n_signature_scheme sig_scheme_out = {0};
+        EXPECT_FAILURE(s2n_choose_tls13_sig_scheme_and_set_cert(server_conn, &empty_client_sig_scheme_pref_list, &sig_scheme_out));
```

Comment:
I'm in favor of defaulting to a preferred cert since this mirrors the behavior of the referenced code. It may be more difficult to reject a client up front without changing the way s2n parses and process handshake.